board/stm32fx/lluart.h

```diff
@@ -0,0 +1,230 @@
+    DMA2_Stream5->M0AR = (uint32_t)q->elems_rx;     // Destination
+    DMA2_Stream5->NDTR = q->rx_fifo_size;           // Number of bytes to copy
+
+    // Circular, Increment memory, byte size, periph -> memory, enable
+    // Transfer complete, half transfer, transfer error and direct mode error interrupt enable
+    DMA2_Stream5->CR = DMA_SxCR_CHSEL_2 | DMA_SxCR_MINC | DMA_SxCR_CIRC | DMA_SxCR_HTIE | DMA_SxCR_TCIE | DMA_SxCR_TEIE | DMA_SxCR_DMEIE | DMA_SxCR_EN;
+
+    // Enable DMA receiver in UART
+    q->uart->CR3 |= USART_CR3_DMAR;
+
+    // Enable UART IDLE interrupt
+    q->uart->CR1 |= USART_CR1_IDLEIE;
+
+    // Enable interrupt
+    NVIC_EnableIRQ(DMA2_Stream5_IRQn);
+  } else {
+    puts("Tried to initialize RX DMA for an unsupported UART\n");
+  }
+}
+
+#define __DIV(_PCLK_, _BAUD_)                    (((_PCLK_) * 25U) / (4U * (_BAUD_)))
+#define __DIVMANT(_PCLK_, _BAUD_)                (__DIV((_PCLK_), (_BAUD_)) / 100U)
+#define __DIVFRAQ(_PCLK_, _BAUD_)                ((((__DIV((_PCLK_), (_BAUD_)) - (__DIVMANT((_PCLK_), (_BAUD_)) * 100U)) * 16U) + 50U) / 100U)
+#define __USART_BRR(_PCLK_, _BAUD_)              ((__DIVMANT((_PCLK_), (_BAUD_)) << 4) | (__DIVFRAQ((_PCLK_), (_BAUD_)) & 0x0FU))
+
+void uart_set_baud(USART_TypeDef *u, unsigned int baud) {
+  if (u == USART1) {
+    // USART1 is on APB2
+    u->BRR = __USART_BRR(48000000U, baud);
+  } else {
+    u->BRR = __USART_BRR(24000000U, baud);
+  }
+}
+
+void uart_init(uart_ring *q, int baud) {
+  // Register interrupts (max data rate: 115200 baud)
+  if(q->uart == USART1){
+    REGISTER_INTERRUPT(USART1_IRQn, USART1_IRQ_Handler, 150000U, FAULT_INTERRUPT_RATE_UART_1)
+  } else if (q->uart == USART2){
+    REGISTER_INTERRUPT(USART2_IRQn, USART2_IRQ_Handler, 150000U, FAULT_INTERRUPT_RATE_UART_2)
+  } else if (q->uart == USART3){
+    REGISTER_INTERRUPT(USART3_IRQn, USART3_IRQ_Handler, 150000U, FAULT_INTERRUPT_RATE_UART_3)
+  } else if (q->uart == UART5){
+    REGISTER_INTERRUPT(UART5_IRQn, UART5_IRQ_Handler, 150000U, FAULT_INTERRUPT_RATE_UART_5)
+  } else {
+    // UART not used. Skip registering interrupts
+  }
+  if(q->dma_rx){
+    REGISTER_INTERRUPT(DMA2_Stream5_IRQn, DMA2_Stream5_IRQ_Handler, 100U, FAULT_INTERRUPT_RATE_UART_DMA)   // Called twice per buffer
+  }
+
+  // Set baud and enable peripheral with TX and RX mode
+  uart_set_baud(q->uart, baud);
+  q->uart->CR1 = USART_CR1_UE | USART_CR1_TE | USART_CR1_RE;
+  if ((q->uart == USART2) || (q->uart == USART3) || (q->uart == UART5)) {
+    q->uart->CR1 |= USART_CR1_RXNEIE;
+  }
+
+  // Enable UART interrupts
+  if(q->uart == USART1){
+    NVIC_EnableIRQ(USART1_IRQn);
+  } else if (q->uart == USART2){
+    NVIC_EnableIRQ(USART2_IRQn);
+  } else if (q->uart == USART3){
+    NVIC_EnableIRQ(USART3_IRQn);
+  } else if (q->uart == UART5){
+    NVIC_EnableIRQ(UART5_IRQn);
+  } else {
+    // UART not used. Skip enabling interrupts
+  }
+
+  // Initialise RX DMA if used
+  if(q->dma_rx){
+    dma_rx_init(q);
+  }
+}
```

Comment:
There is quite a bit of logic in here which I can't believe will be very (if at all) different when implementing on the H7 (stuff like the ring handlers).

Wouldn't it be better to keep this common, but abstract away the specific register differences if there are any? Would like to avoid having the same code twice because they added a bit in a register somewhere...

Reply:
I think we should fully abstract away the hardware from the main panda code then we can abstract the common STMXX stuff on top of that. I think only abstracting away the differences for H7 would leave the code in a weird semi-abstracted state.